crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -1698,86 +1734,31 @@ impl GetFullBatchData for MeshPipeline {
+    fn write_batch_indirect_parameters_metadata(
+        mesh_index: u32,
+        indexed: bool,
+        base_output_index: u32,
+        batch_set_index: Option<NonMaxU32>,
+        indirect_parameters_buffer: &mut IndirectParametersBuffers,
         indirect_parameters_offset: u32,
-        main_entity: MainEntity,
     ) {
-        write_batch_indirect_parameters(
-            mesh_instances,
-            meshes,
-            mesh_allocator,
-            indirect_parameters_buffer,
-            indirect_parameters_offset,
-            main_entity,
-        );
-    }
-}
-
-/// Pushes a set of [`IndirectParameters`] onto the [`IndirectParametersBuffer`]
-/// for the given mesh instance, and returns the index of those indirect
-/// parameters.
-fn write_batch_indirect_parameters(
-    mesh_instances: &RenderMeshInstances,
-    meshes: &RenderAssets<RenderMesh>,
-    mesh_allocator: &MeshAllocator,
-    indirect_parameters_buffer: &mut IndirectParametersBuffer,
-    indirect_parameters_offset: u32,
-    main_entity: MainEntity,
-) {
-    // This should only be called during GPU building.
-    let RenderMeshInstances::GpuBuilding(ref mesh_instances) = *mesh_instances else {
-        error!(
-            "`write_batch_indirect_parameters_index` should never be called in CPU mesh uniform \
-                building mode"
-        );
-        return;
-    };
-
-    let Some(mesh_instance) = mesh_instances.get(&main_entity) else {
-        return;
-    };
-    let Some(mesh) = meshes.get(mesh_instance.mesh_asset_id) else {
-        return;
-    };
-    let Some(vertex_buffer_slice) = mesh_allocator.mesh_vertex_slice(&mesh_instance.mesh_asset_id)
-    else {
-        return;
-    };
-
-    // Note that `IndirectParameters` covers both of these structures, even
-    // though they actually have distinct layouts. See the comment above that
-    // type for more information.
-    let indirect_parameters = match mesh.buffer_info {
-        RenderMeshBufferInfo::Indexed {
-            count: index_count, ..
-        } => {
-            let Some(index_buffer_slice) =
-                mesh_allocator.mesh_index_slice(&mesh_instance.mesh_asset_id)
-            else {
-                return;
-            };
-            IndirectParameters {
-                vertex_or_index_count: index_count,
-                instance_count: 0,
-                first_vertex_or_first_index: index_buffer_slice.range.start,
-                base_vertex_or_first_instance: vertex_buffer_slice.range.start,
-                first_instance: 0,
-            }
-        }
-        RenderMeshBufferInfo::NonIndexed => IndirectParameters {
-            vertex_or_index_count: mesh.vertex_count,
+        let indirect_parameters = IndirectParametersMetadata {
+            mesh_index,
+            base_output_index,
+            batch_set_index: match batch_set_index {
+                Some(batch_set_index) => u32::from(batch_set_index),
+                None => !0,
```

Comment:
Why `!0`?

Reply:
Line 79 of `build_indirect_params.wgsl` checks this value to see whether the batch belongs to a batch set.